File: agents/buyer_agent.py
import json
from web3 import Web3

class PythonBuyerAgent:
    def __init__(self, w3, signer_account, contract_address, abi):
        self.w3 = w3
        self.account = signer_account
        self.contract = w3.eth.contract(address=contract_address, abi=abi)

    def evaluate_listing(self, listing_id):
        print(f"[Python Buyer Agent] Reading reputation & listing #{listing_id} on-chain...")
        eval_data = self.contract.functions.evaluateListing(listing_id).call()
        is_valid, can_purchase, seller_score, seller_stake, price = eval_data

        listing = self.contract.functions.listings(listing_id).call()
        seller_rep = self.contract.functions.getSellerReputation(listing[1]).call()

        print(f"[Python Buyer Agent] Teaser: '{listing[3]}'")
        print(f"[Python Buyer Agent] Price: {self.w3.from_wei(price, 'ether')} ETH")
        print(f"[Python Buyer Agent] Seller Stake: {self.w3.from_wei(seller_stake, 'ether')} ETH")
        print(f"[Python Buyer Agent] Seller Reputation Score: {seller_score} bps ({seller_score/100:.1f}%)")

        min_stake = self.w3.to_wei(0.01, 'ether')
        min_score = 2500

        if is_valid and can_purchase and seller_stake >= min_stake and seller_score >= min_score:
            print("[Python Buyer Agent] DECISION: APPROVED for Escrow Purchase!")
            return True, price, listing[2] # return should_buy, price, commitment
        else:
            print("[Python Buyer Agent] DECISION: REJECTED due to high risk / low stake")
            return False, price, listing[2]

    def buy_intel(self, listing_id, price_wei):
        print(f"[Python Buyer Agent] Locking payment into escrow for Listing #{listing_id}...")
        tx = self.contract.functions.buyIntel(listing_id).build_transaction({
            'from': self.account.address,
            'value': price_wei,
            'nonce': self.w3.eth.get_transaction_count(self.account.address),
            'gas': 300000,
            'gasPrice': self.w3.eth.gas_price
        })
        signed = self.account.sign_transaction(tx)
        tx_hash = self.w3.eth.send_raw_transaction(signed.raw_transaction)
        self.w3.eth.wait_for_transaction_receipt(tx_hash)
        print(f"[Python Buyer Agent] Payment locked in escrow contract. Tx: {tx_hash.hex()}")
        return tx_hash.hex()

    def confirm_receipt(self, purchase_id):
        print(f"[Python Buyer Agent] Confirming receipt for Purchase #{purchase_id}...")
        tx = self.contract.functions.confirmReceipt(purchase_id).build_transaction({
            'from': self.account.address,
            'nonce': self.w3.eth.get_transaction_count(self.account.address),
            'gas': 200000,
            'gasPrice': self.w3.eth.gas_price
        })
        signed = self.account.sign_transaction(tx)
        tx_hash = self.w3.eth.send_raw_transaction(signed.raw_transaction)
        self.w3.eth.wait_for_transaction_receipt(tx_hash)
        print(f"[Python Buyer Agent] Escrow released to seller. Tx: {tx_hash.hex()}")
        return tx_hash.hex()

File: agents/buyer_agent.js
const { ethers } = require("ethers");

class BuyerAgent {
  constructor(signer, contractAddress, contractAbi) {
    this.signer = signer;
    this.contract = new ethers.Contract(contractAddress, contractAbi, signer);
  }

  async evaluateListing(listingId) {
    console.log(`[Buyer Agent] Evaluating Listing #${listingId} on-chain...`);
    const evalData = await this.contract.evaluateListing(listingId);

    const [isValid, canPurchase, sellerScoreBps, sellerStake, price] = evalData;

    const listing = await this.contract.listings(listingId);
    const sellerRep = await this.contract.getSellerReputation(listing.seller);

    console.log(`[Buyer Agent] Teaser: "${listing.teaser}"`);
    console.log(`[Buyer Agent] Price: ${ethers.formatEther(price)} ETH`);
    console.log(`[Buyer Agent] Seller Stake: ${ethers.formatEther(sellerStake)} ETH`);
    console.log(`[Buyer Agent] Seller Score: ${sellerScoreBps.toString()} bps (${(Number(sellerScoreBps)/100).toFixed(1)}%)`);
    console.log(`[Buyer Agent] Seller Sales Track Record: ${sellerRep.completedSales.toString()} completed sales`);

    // Decision Logic: Blind Purchase Risk Assessment
    const minStakeWei = ethers.parseEther("0.01");
    const minScoreBps = 2500; // Minimum 25% reputation threshold for trial purchase

    let decision = {
      shouldBuy: false,
      reason: ""
    };

    if (!isValid || !canPurchase) {
      decision.reason = "Listing is invalid or expired";
    } else if (sellerStake < minStakeWei) {
      decision.reason = "Seller has insufficient stake locked (lacks skin in the game)";
    } else if (sellerScoreBps < minScoreBps) {
      decision.reason = `Seller reputation score (${sellerScoreBps}) is below minimum threshold (${minScoreBps})`;
    } else {
      decision.shouldBuy = true;
      decision.reason = `Passed risk check: Stake (${ethers.formatEther(sellerStake)} ETH >= 0.01 ETH), Score (${sellerScoreBps} bps >= ${minScoreBps} bps)`;
    }

    console.log(`[Buyer Agent] Decision Outcome: ${decision.shouldBuy ? "APPROVED" : "REJECTED"} -> ${decision.reason}`);
    return { ...decision, price, seller: listing.seller, commitment: listing.contentCommitment };
  }

  async purchaseIntel(listingId, priceWei) {
    console.log(`[Buyer Agent] Initiating Escrow Purchase for Listing #${listingId}...`);
    const tx = await this.contract.buyIntel(listingId, { value: priceWei });
    const receipt = await tx.wait();

    let purchaseId = 1;
    for (const log of receipt.logs) {
      try {
        const parsed = this.contract.interface.parseLog(log);
        if (parsed && parsed.name === "IntelPurchased") {
          purchaseId = parsed.args.purchaseId;
          break;
        }
      } catch (e) {}
    }

    console.log(`[Buyer Agent] Payment of ${ethers.formatEther(priceWei)} ETH locked in Escrow! Purchase #${purchaseId}. Tx: ${tx.hash}`);
    return { purchaseId, txHash: tx.hash };
  }

  async verifyAndConfirm(purchaseId, expectedCommitment) {
    console.log(`[Buyer Agent] Inspecting revealed content for Purchase #${purchaseId}...`);
    const purchase = await this.contract.purchases(purchaseId);

    if (purchase.status !== 1n && purchase.status !== 1) { // PurchaseStatus.Revealed
      throw new Error(`Purchase status is not Revealed (status: ${purchase.status})`);
    }

    const revealedText = purchase.revealedContent;
    console.log(`[Buyer Agent] Received Revealed Content: "${revealedText}"`);

    // Verify hash commitment on-chain / off-chain
    const computedHash = ethers.keccak256(ethers.toUtf8Bytes(revealedText));
    if (computedHash !== expectedCommitment) {
      console.log(`[Buyer Agent] ALERT: Hash mismatch! Expected ${expectedCommitment}, got ${computedHash}`);
      return { verified: false, revealedText };
    }

    console.log(`[Buyer Agent] Hash match verified! Confirming receipt and releasing escrow...`);
    const tx = await this.contract.confirmReceipt(purchaseId);
    await tx.wait();
    console.log(`[Buyer Agent] Escrow released to seller! Tx: ${tx.hash}`);
    return { verified: true, revealedText, txHash: tx.hash };
  }

  async raiseDispute(purchaseId, reason, disputeBondEth = "0.001") {
    console.log(`[Buyer Agent] Raising dispute for Purchase #${purchaseId}. Reason: "${reason}"`);
    const disputeBondWei = ethers.parseEther(disputeBondEth);
    const tx = await this.contract.disputePurchase(purchaseId, reason, { value: disputeBondWei });
    await tx.wait();
    console.log(`[Buyer Agent] Dispute filed on-chain! Tx: ${tx.hash}`);
    return tx.hash;
  }
}

module.exports = BuyerAgent;
